src/frontend/src/helpers/storage.ts

```diff
@@ -0,0 +1,416 @@
+import Dexie from 'dexie';
+import 'jquery.cookie';
+import {md5} from 'blueimp-md5';
+import {sprintf} from 'sprintf-js';
+
+// need to import SEASHELL_CREDS_COOKIE
+var SEASHELL_CREDS_COOKIE = "";
+// need to import SEASHELL_CREDS_COOKIE
+var USERNAME = "";
+
+class StorageDB extends Dexie {
+  public changelog: Dexie.Table<ChangeLog, number>;
+  public files: Dexie.Table<File, string[]>;
+  public projects: Dexie.Table<Project, string>;
+  public settings: Dexie.Table<Settings, string>;
+
+  public constructor(location: string) {
+    super(location);
+    this.version(1).stores({
+      changelog: '++id',
+      files: '[project+file], project',
+      projects: 'name',
+      settings: 'name'
+    });
+  }
+}
+
+// subject to changes
+interface ChangeLog {
+}
+
+// subject to changes
+interface File {
+  project: string;
+  name: string;
+  contents: string;
+  history: string;
+  checksum: string;
+  lastModified: Date;
+}
+
+// subject to changes
+interface Project {
+  settings: any;
+  name: string;
+}
+
+// subject to changes
+interface Settings {
+
+}
+
+export class Storage {
+
+  private db: StorageDB;
+  private username: string;
+  private has_offline_changes: boolean;
+
+  public constructor() {
+    this.db = new StorageDB(USERNAME + "/seashell-local-files");
+
+    this.username = $.cookie(SEASHELL_CREDS_COOKIE).user;
+    this.has_offline_changes = false;
+  }
+
+  /*
+    * Save a file to local storage.
+    * @param {string} name: project name.
+    * @param {string} file_name: filename.
+    * @param {string} file_content: The contents of the file.
+    * @param {string} file_history: The history of the file.
+    * @param {string || any false value} checksum : The online checksum of the file, false to not update (offline write).
+    */
+  public writeFile(file: string, contents: string, history: string, checksum: string) {
+    var self = this;
+    var offline_checksum = md5(contents);
+    var key = [name, file];
+    return self.db.transaction('rw', self.db.changelog, self.db.files, function () {
+      if (checksum !== undefined) {
+        return self.db.files.update(key, {
+          contents: contents, 
+          history: history, 
+          checksum: checksum,
+          last_modified: Date.now()
+        }).then(function (result) {
+          if (! result) {
+            throw sprintf("Storage.writeFile: file %s/%s not found!", name, file);
```

Comment:
We shouldn't actually throw strings.  It would be nice if you could create a class for Storage errors and throw that instead.